CASSANDRA-16914. Implement Virtual Tables for Auth Caches.

azotcsit opened the pull request:
It is raised for self-review. I'll convert it to a proper PR once all the changes are ready.

azotcsit committed:
Implement Virtual Tables for Auth Caches

Patch by Aleksei Zotov; reviewed by Sam Tunnicliffe, and Benjamin Lerer for CASSANDRA-16914

azotcsit commented on src/java/org/apache/cassandra/auth/Roles.java:
```diff
@@ -38,26 +35,8 @@
-    {
-        cache.invalidate();
-    }
+    public static final RolesCache cache = new RolesCache(DatabaseDescriptor.getRoleManager(),
```
Similarly to other caches it is setup in tests during initialization instead of changing the behavior via "VisibleForTesting methods". Here I ensured the cache instance is immutable (stored in a static final variable), so it is safe to expose.

beobal replied:
This seems to have overlapped with CASSANDRA-17016 being committed and which makes use of `initRolesCache`.

azotcsit replied:
Good catch! Again rebase did not highlight the issue, I'll make sure to run the build locally next time. Fixed now.

azotcsit commented on src/java/org/apache/cassandra/auth/PasswordAuthenticator.java:
```diff
@@ -69,14 +69,18 @@
         return true;
     }
 
+    public static CredentialsCache getCredentialsCache() {
```
Need to discuss this and decided how we want to deal with this. From one hand, it exposes the cache outside and a kind of breaks encapsulation. From another hand, I really need the access to the cache from outside. Theoretically I could expose some particular methods, but:
1. There are many methods to expose
- invalidate()
- invalidate(key)
- getAll()
- get(key)
2. exposing some cache-related methods from PasswordAuthenticator/Roles/etc classes seems to be a bit confusing

The same problem is actual for ALL caches. For now, I made all of them public, so I could access them. But we need to decided what we want to do - keep it as is / expose methods partially / make some structural changes (not sure what exactly).

The current implementation with public access to the static caches still seems to be same since the caches are thread-safe and immutable (I mean the static instance variables - aka constants).

I'd be happy to hear any suggestions.

beobal replied:
I don't see a problem with making the caches public and breaking encapsulation/making them all first class citizens. In fact, they are intended to be independent of the underlying auth implementations as far as is possible, it's just that the nature of SASL authentication doesn't lend itself to a generic caching approach in the same way as simple user/pass credentials do.

azotcsit commented on src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java:
```diff
@@ -145,7 +145,7 @@ protected void applyRangeTombstone(ColumnValues partitionKey, Range<ColumnValues
     }
 
-    protected void applyRowDeletion(ColumnValues partitionKey, ColumnValues clusteringColumnValues)
+    protected void applyRowDeletion(ColumnValues partitionKey, ColumnValues clusteringColumns)
```
Missed that on review, so fixing now.

azotcsit commented on test/unit/org/apache/cassandra/auth/RolesTest.java:
```diff
@@ -40,14 +46,12 @@
-        IRoleManager roleManager = new LocalCassandraRoleManager();
-        roleManager.setup();
-        Roles.initRolesCache(roleManager, () -> true);
+        LocalCassandraRoleManager roleManager = new LocalCassandraRoleManager();
```
That's the right way to initialize this test, so there won't be a need for `Roles.initRolesCache` method.

azotcsit commented on doc/source/new/virtualtables.rst:
```diff
@@ -308,6 +324,56 @@ As another example, to find how much time is remaining for SSTable tasks, use th
+
+::
+
+  cqlsh:system_views> select * from credentials_cache;
```
I have a general concern on the current structure of this page - it is hard to maintain it (keep up-to-date). Maybe we should not list all table here? It won't be a part of this ticket for sure, I just want to raise the question...

beobal commented on src/java/org/apache/cassandra/db/virtual/JmxPermissionsCacheTable.java:
```diff
@@ -0,0 +1,94 @@
+                .addPartitionKeyColumn(ROLE, UTF8Type.instance)
+                .addClusteringColumn(GRANTEE, UTF8Type.instance)
+                .addClusteringColumn(RESOURCE, UTF8Type.instance)
+                .addClusteringColumn(PERMISSION, UTF8Type.instance)
```
Same comment applies here as to the main (non-jmx) permissions cache wrt to granularity & clustering columns

azotcsit replied:
This class does not exist anymore. The latest version has `.addPartitionKeyColumn(ROLE, UTF8Type.instance)` only.

It seems to be an old comment that you added to review some time ago and posted now.

beobal replied:
hmm, sorry about that. I did start the review before your last update, but when I submitted it it didn't show me any of the outdated comments. Sorry for the noise!

azotcsit commented on src/java/org/apache/cassandra/tools/nodetool/InvalidatePermissionsCache.java:
```diff
@@ -38,7 +38,7 @@
 @Command(name = "invalidatepermissionscache", description = "Invalidate the permissions cache")
 public class InvalidatePermissionsCache extends NodeToolCmd
 {
-    @Arguments(usage = "[<user>]", description = "A specific user for whom permissions need to be invalidated")
```
I renamed user to role here to be coherent with the VT. Also "user" term seems to be outdated since everything relies onto roles now.

beobal commented on test/unit/org/apache/cassandra/auth/CassandraNetworkAuthorizerTest.java:
```diff
@@ -63,13 +62,12 @@ public static void setupSuperUser()
-                               new PasswordAuthenticator(),
-                               new AuthTestUtils.LocalCassandraAuthorizer(),
-                               new AuthTestUtils.LocalCassandraNetworkAuthorizer());
+        SchemaLoader.setupAuth(new AuthTestUtils.LocalCassandraRoleManager(),
```
nit: this is highly subjective, but I find the original alignment more readable. It wouldn't stop me +1'ing a patch, but I just wanted to mention in case it was done by an IDE unintentionally. (nb. applies in multiple locations in the PR)

azotcsit replied:
Yes, it was automatically formatted by Intellij. I fixed it everywhere, but generally I'm not sure how to deal with this problem. We either need to document some project-wide stylistic preferences or define some checkstyle rules. Otherwise, people will keep changing formatting based on their subjective preferences.

beobal replied:
There is some documentation on this [here](https://cassandra.apache.org/_/development/code_style.html), although when I went to check it, it appears to have gotten a bit mangled over the course of a few site updates. 

The alignment as illustrated in the style guide looks more like yours/Intellij's, but previously this was specified in the way it was initially [(see history)](https://web.archive.org/web/20191020015442/http://cassandra.apache.org/doc/latest/development/code_style.html)

The code style page does include links to style configs for Eclipse and Intellij, but I don't know whether these are compatible with the latest versions (the link to the Intellij jar is mostly broken anyway, requiring a few extra clicks). I'll open a Jira / PR to update the docs.

azotcsit replied:
Oh ok, by some reason I did not see the Intellij config gist. I'll definitely give it a try. Thanks for referencing!

beobal commented on test/unit/org/apache/cassandra/db/virtual/CredentialsCacheKeysTableTest.java:
```diff
@@ -0,0 +1,180 @@
+                new AuthTestUtils.LocalCassandraAuthorizer(),
+                new AuthTestUtils.LocalCassandraNetworkAuthorizer());
+
+        roleManager.createRole(AuthenticatedUser.SYSTEM_USER, ROLE_A, AuthTestUtils.getLoginRoleOprions());
```
typo here, which is common to all the new `*CacheKeysTableTest` and causes compilation failures.

azotcsit replied:
Oh ok, it is a rebase issue. After rebasing onto the latest trunk the method name got fixed (CASSANDRA-16926), but since there was no conflict here, I missed it. Fixed everywhere.

beobal reviewed (changes requested):
Mostly LGTM but there are a handful of typos which cause build failures. I've also got  one minor style nit, but I'll leave it to your discretion whether to fix it or not. 
I think we need another CI run too.

azotcsit commented:
@beobal 
By some reason I cannot reply to https://github.com/apache/cassandra/pull/1208#discussion_r729746327. You're correct. It is another issue appeared after rebase which I missed. So I'll be running local build for every rebase (even without conflicts) next time.

I fixed all the comments and started CI: https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch/1208/. Let's wait for the results.
**UPDATE:** I checked the failures and they do not seem to be related to these changes.

beobal reviewed (changes requested):
Sorry, this is the last round I'm sure...

In `AbstractMutableVirtualTable`, the import of `org.apache.commons.lang.ArrayUtils` needs updating to the `lang3` version. As it is, the build and tests are fine due to the older `commons-lang` jar being available on account of it being a transitive dependency of Hadoop. When actually running C* itself, either from the source tree or the binary distro, Hadoop and its deps aren't on the classpath so we get a `NoClassDefFoundError` when deleting from the vtables.

One other related thing I noticed while testing this is that `nodetool invalidatepermissionscache <role>` has no effect. This is expected as the cache keys are (role, resource) tuples, but it might be helpful if we made it mandatory to supply a resource too. Sorry I missed this in the CASSANDRA-16404 review.

azotcsit commented:
> Sorry, this is the last round I'm sure...

No worries! More rounds we have better the results are :+1: 

> When actually running C* itself, either from the source tree or the binary distro, Hadoop and its deps aren't on the classpath so we get a NoClassDefFoundError when deleting from the vtables.

Oh.. it is a pity that the build does not guarantee correctness of the code.. I think once we migrate from `ant`, it is worth to consider moving Hadoop stuff to a separate module. Anyway, it is too early to think about it. Great catch! Done.

> One other related thing I noticed while testing this is that nodetool invalidatepermissionscache <role> has no effect.

Oh! I fixed it and added a corresponding test. Thanks for spotting the issue!


PS: I triggered a new CI build for the latest changes: https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch/1223/ 
**UPDATE:** CI looks good

azotcsit commented:
I also pushed a commit with small fixes for logic introduced by CASSANDRA-17027.

blerer commented on src/java/org/apache/cassandra/auth/PasswordAuthenticator.java:
```diff
@@ -78,6 +78,10 @@ public boolean requireAuthentication()
         return true;
     }
 
+    public CredentialsCache getCredentialsCache() {
```
nit: The brace should be on the next line

azotcsit replied:
Fixed.

blerer commented on src/java/org/apache/cassandra/db/virtual/CredentialsCacheKeysTable.java:
```diff
@@ -0,0 +1,70 @@
+        IAuthenticator authenticator = DatabaseDescriptor.getAuthenticator();
+        if (authenticator instanceof PasswordAuthenticator)
+            ((PasswordAuthenticator) authenticator).getCredentialsCache().getAll()
+                .forEach((roleName, ignored) -> result.row(roleName));
```
The logic to retrieve the credential cache could be extracted. Moreover, if I am not mistaken the IAuthenticator cannot be changed without restarting the server so we could retrieve the Credential cache only once in the constructor.

azotcsit replied:
I double checked whether `IAuthenticator` is not supposed to be changed in runtime - that's correct. So I started setting up it in constructor.

blerer commented on test/unit/org/apache/cassandra/service/ClientStateTest.java:
```diff
@@ -64,7 +62,7 @@ public static void afterClass()
 
     @Test
-    public void permissionsCheckStartsAtHeadOfResourceChain() throws Exception {
+    public void permissionsCheckStartsAtHeadOfResourceChain() {
```
nit: The brace should be on the next line.

azotcsit replied:
Fixed.

blerer commented on test/unit/org/apache/cassandra/db/virtual/PermissionsCacheKeysTableTest.java:
```diff
@@ -0,0 +1,196 @@
+        DatabaseDescriptor.setPermissionsValidity(20_000);
+
+        CQLTester.setUpClass();
+        AuthTestUtils.LocalCassandraRoleManager roleManager = new AuthTestUtils.LocalCassandraRoleManager();
```
`CQLTester.requireAuthentication` can be used to setup the authentication and automatically create the super user. After that it is possible to use CQL queries to create roles and grant/revoke permissions

azotcsit replied:
I changed tests to use `CQLTester.requireAuthentication` for local auth setup. However, I did not migrate them to use `useSuperUser`/`useUser` because these methods cannot be referred from static `@BeforeClass`. Theoretically I could put them into `@Before/@After` but I felt it is an overhead (not a problem though) to re-create resources for every test while it could be done in static `@BeforeClass` once and for all.

Another advantage in the current implementation is that it is possible to easily use `IResource.applicablePermissions`. For CQL version it would require more code to construct statements and look cumbersome.

Please, let me know your thoughts. If you believe that CQL is a better way, I'd like to understand the advantages. The only I can see is less dependencies on internal Java implementation (no need to changes tests if some Java classes are changed).

blerer commented on test/unit/org/apache/cassandra/db/virtual/CredentialsCacheKeysTableTest.java:
```diff
@@ -0,0 +1,180 @@
+    @Test
+    public void testSelectAllWhenPermissionsAreNotCached() throws Throwable
+    {
+        UntypedResultSet result = execute("SELECT * FROM vts.credentials_cache_keys");
```
The practice within the C* unit test is to inline the execute within the assertion. It is often easier to read to see the query next to the results.

azotcsit replied:
Done for all tests.
